File: amazon/node/amazon-medication-prices-by-condition.js
//const axios = require('axios');
import axios from 'axios';
import * as cheerio from 'cheerio';
import { createObjectCsvWriter } from 'csv-writer'; 
//const createCsvWriter = require('csv-writer').createObjectCsvWriter;
// const mysql = require('mysql2'); // Scaffold for later SQL use

const START_URL = 'https://www.amazon.com/s?k=high+blood+pressure&i=amazon-pharmacy&ref=sf_highbloodpressure';

const csvWriter = createObjectCsvWriter({
  path: 'drug_data.csv',
  header: [
    { id: 'drugName', title: 'Drug Name' },
    { id: 'form', title: 'Form' },
    { id: 'strength', title: 'Strength' },
    { id: 'no_insurance_price', title: 'Price w/o Insurance' },
    { id: 'insurance_price', title: 'Price w/ Insurance' },
    { id: 'supply', title: 'Supply (Days)' },
    { id: 'url', title: 'URL' },
  ]
});

async function fetchHTML(url) {
  const { data } = await axios.get(url, {
    headers: {
      'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36'
    }
  });
   
  return cheerio.load(data);
}

async function extractDrugLinks($) {
  const links = [];
  $('a.a-link-normal.s-no-outline').each((_, el) => {
    const href = $(el).attr('href');
    if (href && href.includes('/dp/')) {
      //const fullUrl = `https://www.amazon.com${href}`;
      const fullUrl = `${href}`;
      links.push(fullUrl);
    }
  });
  return [...new Set(links)];
}

async function extractDrugDetails(url) {
  try {
    const $ = await fetchHTML(url);

    const drugName = $('h1.pui-heading').first().text().trim();
    //const form = $('span:contains("Form")').next().text().trim() || '';
    //const strength = $('span:contains("Strength")').next().text().trim() || '';
    //const price = $('#corePrice_feature_div span.a-price span.a-offscreen').first().text().trim() || '';
    //const supplyMatch = $('span:contains("supply")').text().match(/(\d+)[ -]?day/i);
    //const supply = supplyMatch ? supplyMatch[1] : '';


    const form = $('#form-box-item').text().trim();
    const strength = $('#strength-box-item').text().trim();
    const frequency = $('#frequency-box-item').text().trim();
    const supply = $('#supply-box-item').text().trim();

    // Pricing info
    const no_insurance_price = $('#insurance-estimate-median-price').text().trim();
    console.log('Median Price:', no_insurance_price);
    const insurance_price = $('#extended-supply-price-label').text().trim();
  
    console.log('Drug Name:', drugName);
    console.log('Form:', form);
    console.log('Strength:', strength);
    console.log('Frequency:', frequency);
    console.log('Supply:', supply);

    return { drugName, form, strength, no_insurance_price, insurance_price, supply, url };
  } catch (err) {
    console.error(`Error scraping ${url}:`, err.message);
    return null;
  }
}

(async () => {
  try {
    const $ = await fetchHTML(START_URL);
    const drugLinks = await extractDrugLinks($);
    console.log(`Found ${drugLinks.length} drug links.`);

    console.log(drugLinks);

    const results = [];

    for (const link of drugLinks) {
      const data = await extractDrugDetails(link);
      if (data) {
        results.push(data);
        console.log(`Scraped: ${data.name}`);
      }
      await new Promise(res => setTimeout(res, 1000)); // polite delay
    }

    await csvWriter.writeRecords(results);
    console.log(`Finished writing ${results.length} entries to drug_data.csv`);

    // Scaffold for future SQL connection
    /*
    const connection = mysql.createConnection({
      host: 'localhost',
      user: 'your_user',
      password: 'your_password',
      database: 'your_database'
    });

    connection.connect();

    for (const drug of results) {
      const query = `INSERT INTO Drugs (Name, Form, Strength, Price, Supply, Url)
                     VALUES (?, ?, ?, ?, ?, ?)`;
      const values = [drug.name, drug.form, drug.strength, drug.price, drug.supply, drug.url];
      connection.query(query, values, (err) => {
        if (err) console.error('Insert error:', err.message);
      });
    }

    connection.end();
    */

  } catch (err) {
    console.error('Error:', err.message);
  }
})();

File: amazon/amazon-medication-prices-by-condition.py
import csv
import time
from datetime import datetime
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from webdriver_manager.chrome import ChromeDriverManager

# Set up headless Chrome
options = webdriver.ChromeOptions()
options.add_argument('--headless')
options.add_argument('--disable-gpu')
options.add_argument('--no-sandbox')
driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=options)

# URL to Amazon Pharmacy search results
#search_url = "https://www.amazon.com/s?k=high+blood+pressure&i=amazon-pharmacy&ref=sf_highbloodpressure"
#search_url = "https://www.amazon.com/s?k=high+cholesterol&i=amazon-pharmacy&ref=sf_highcholesterol"
#search_url = "https://www.amazon.com/s?k=depression&i=amazon-pharmacy&ref=sf_depression"
search_url = "https://www.amazon.com/s?k=anxiety&i=amazon-pharmacy&ref=sf_anxiety"
#search_url = "https://www.amazon.com/s?k=acid+reflux&i=amazon-pharmacy&ref=sf_acidreflux"
#search_url = "https://www.amazon.com/s?k=birth+control&i=amazon-pharmacy&ref=sf_birthcontrol"
#search_url = "https://www.amazon.com/s?k=diabetes&i=amazon-pharmacy&ref=sf_diabetes"
#search_url = "https://www.amazon.com/s?k=hypothyroidism&i=amazon-pharmacy&ref=sf_hypothyroidism"
#search_url = "https://www.amazon.com/s?k=allergies&i=amazon-pharmacy&ref=sf_allergies"
#search_url = "https://www.amazon.com/s?k=asthma&i=amazon-pharmacy&ref=sf_asthma"
#search_url = "https://www.amazon.com/s?k=hair+loss&i=amazon-pharmacy&ref=sf_hairloss"

category = search_url.split('_')[-1]
driver.get(search_url)
time.sleep(3)

# Step 1: Collect drug detail links
drug_links = []
product_elements = driver.find_elements(By.XPATH, "//a[@class='a-link-normal s-no-outline']")
for elem in product_elements:
    href = elem.get_attribute("href")
    if href and "amazon.com" in href:
        drug_links.append(href)

print(f"Collected {len(drug_links)} drug links")

# Step 2: Visit each drug link and collect information
drug_data = []

for link in drug_links[:100]:  # Limit for demonstration
    if ".com/r?aid" not in link:
        driver.get(link)
        time.sleep(3)

        # Basic Info
        try:
            drug_name = driver.find_element(By.TAG_NAME, "h1").text.strip()
        except:
            drug_name = "N/A"

        def try_get_by_xpath(xpath):
            try:
                return driver.find_element(By.XPATH, xpath).text.strip()
            except:
                return "N/A"

        def try_get_by_id(element_id):
            try:
                return driver.find_element(By.ID, element_id).text.strip()
            except:
                return "N/A"


        def format_price(price):
            try:
                price = price.replace("\n", "").replace("$", "")
                price_int = int(price)
                price_float = float(price_int / 100)
                formatted_price = f"{price_float:.2f}"
                print(formatted_price)
                return formatted_price        
            except:
                return ""

        form = try_get_by_id("form-box-item")
        strength = try_get_by_id("strength-box-item")
        frequency = try_get_by_id("frequency-box-item")
        supply = try_get_by_id("supply-box-item")
        
        # Pricing Info
        no_insurance_price = try_get_by_id("insurance-estimate-median-price")
        insurance_price = try_get_by_id("extended-supply-price-label")  # spelling from your message
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        
        print(drug_name)
        no_insurance_price_formatted = format_price(no_insurance_price)
        insurance_price_formatted = format_price(insurance_price)
        link = link.split('?')[0]

        drug_data.append({
            "Drug Name": drug_name,
            "Price URL": link,
            "Form": form,
            "Strength": strength,
            "Frequency": frequency,
            "Supply": supply,
            "Average Insurance Price": insurance_price_formatted,
            "Buy Without Insurance": no_insurance_price_formatted,
            "Time Stamp": timestamp
        })

# Step 3: Save to CSV
csv_filename = f"amazon_drugs_{category}.csv"
with open(csv_filename, mode='w', newline='', encoding='utf-8') as file:
    writer = csv.DictWriter(file, fieldnames=["Drug Name", "Price URL", "Form", "Strength", "Frequency", "Supply", "Average Insurance Price", "Buy Without Insurance", "Time Stamp"])
    writer.writeheader()
    writer.writerows(drug_data)

print(f"\nSaved {len(drug_data)} drug entries to {csv_filename}")
driver.quit()
